components/servo/message_port.rs

```diff
@@ -0,0 +1,163 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use base::id::{MessagePortId, MessagePortRouterId, WebViewId};
+use constellation_traits::{EmbedderToConstellationMessage, MessagePortMsg, PostMessageData};
+use embedder_traits::{EmbedderMsg, EmbedderProxy, JSValue};
+use ipc_channel::ipc;
+use ipc_channel::router::ROUTER;
+
+use crate::{ConstellationProxy, WebView};
+
+pub trait MessagePortDelegate {
+    /// Invoked when a message is posted on an entangled MessagePort. The
+    /// posted data is provided as an argument, and only includes seralized
+    /// versions of fundamental JS data types.
+    fn post_message_received(&self, _webview: WebView, _message_port: MessagePort, _data: JSValue) {
+    }
+}
+
+enum MessagePortState {
+    Ready {
+        id: MessagePortId,
+        webview_id: WebViewId,
+        entangled_id: MessagePortId,
+        constellation_proxy: ConstellationProxy,
+        delegate: Option<Rc<dyn MessagePortDelegate>>,
+    },
+    Detached,
+}
+
+#[derive(Clone)]
+pub struct MessagePort(Rc<RefCell<MessagePortState>>);
+
+#[derive(Debug, PartialEq)]
+pub struct MessagePortDetached;
+
+impl MessagePort {
+    fn new(
+        constellation_proxy: ConstellationProxy,
+        webview_id: WebViewId,
+        port_id: MessagePortId,
+        entangled_id: MessagePortId,
+    ) -> MessagePort {
+        MessagePort(Rc::new(RefCell::new(MessagePortState::Ready {
+            constellation_proxy,
+            webview_id,
+            id: port_id,
+            entangled_id,
+            delegate: None,
+        })))
+    }
+
+    pub(crate) fn new_entangled(
+        constellation_proxy: ConstellationProxy,
+        embedder_proxy: EmbedderProxy,
+        webview_id: WebViewId,
+    ) -> (MessagePort, MessagePort) {
+        let (port_control_sender, port_control_receiver) =
+            ipc::channel().expect("ipc channel failure");
+        ROUTER.add_typed_route(
```

Comment:
Could we use the `GenericCallback` mechanism here to abstract the multi-process boundary?